JSON-Grammar-spec.md

```diff
@@ -0,0 +1,794 @@
+
+**Shorthand Array:**
+```json
+["{", {"r": "Content"}, "}"]
+```
+
+**Full Production Node:**
+```json
+{
+  "p": ["{", {"r": "Content"}, "}"],
+  "type": "Block",
+  "children": {
+    "1": "content"
+  }
+}
+```
+
+#### Examples
+
+```json
+{
+  "cst": {
+    "FunctionCall": ["func", "(", ")"],
+    "Assignment": {
+      "p": [{"r": "Identifier"}, "=", {"r": "Expression"}],
+      "type": "Assignment",
+      "children": {
+        "0": "target",
+        "2": "value"
+      }
+    },
+    "IfStatement": {
+      "p": ["if", "(", {"r": "Expression"}, ")", {"r": "Statement"}],
+      "children": {
+        "2": "condition",
+        "4": "body"
+      }
+    }
+  }
+}
+```
+
+### 4. UnionNode (Union Node)
+
+Matches one of several alternative patterns. The first matching alternative is selected (ordered choice).
+
+#### Interface
+```typescript
+interface UnionNode {
+  u: GrammarNode[];           // Array of alternative nodes
+  type?: string;              // Type name (default: "Union")
+  ast?: AstNodeExpression;    // AST transformation
+}
+```
+
+#### Syntax
+```json
+{
+  "u": ["pattern1", "pattern2", "pattern3"]
+}
+```
+
+#### Examples
+
+```json
+{
+  "cst": {
+    "Literal": {
+      "u": ["null", "true", "false", {"r": "Number"}, {"r": "String"}]
+    },
+    "Statement": {
+      "u": [
+        {"r": "IfStatement"},
+        {"r": "ReturnStatement"},
+        {"r": "ExpressionStatement"}
+      ]
+    },
+    "BinaryOperator": {
+      "u": ["+", "-", "*", "/", "==", "!=", "<", ">"]
+    }
+  }
+}
+```
+
+### 5. ListNode (List Node)
+
+Matches zero or more repetitions of a pattern.
+
+#### Interface
+```typescript
+interface ListNode {
+  l: GrammarNode;              // Node to repeat
+  type?: string;               // Type name (default: "List")
+  ast?: AstNodeExpression;     // AST transformation
+}
+```
+
+#### Syntax
+```json
+{
+  "l": "pattern"
+}
+```
+
+#### Examples
+
+```json
+{
+  "cst": {
+    "Statements": {
+      "l": {"r": "Statement"}
+    },
+    "Parameters": {
+      "l": {
+        "p": [",", {"r": "Parameter"}],
+        "ast": ["$", "/children/1"]
+      }
+    },
+    "Digits": {
+      "l": "/[0-9]/",
+      "type": "DigitSequence"
+    }
+  }
+}
+```
+
+## AST Conversion
+
+The Abstract Syntax Tree (AST) conversion process transforms the detailed Concrete Syntax Tree (CST) into a simplified, semantically meaningful tree structure suitable for further processing.
+
+### Default AST Generation
+
+When no custom AST transformation is specified, the library generates a canonical AST node:
+
+```typescript
+interface CanonicalAstNode {
+  type: string;                                    // Node type
+  pos: number;                                     // Start position
+  end: number;                                     // End position
+  raw?: string;                                    // Raw matched text
+  children?: (CanonicalAstNode | unknown)[];      // Child nodes
+}
+```
+
+### AST Expression System
```

Comment:
Before mentioning AST Expression System, mention the simplified AST construction method using `children` syntax.